feat(service-worker): improve ergonomics of the `SwUpdate` APIs

gkalpak opened the pull request:
_NOTE: This is cleaned up version of #43535._

##
See the original PR and individual commits for details.

Closes #43535

mary-poppins commented:
You can preview 49a920c at https://pr43668-49a920c.ngbuilds.io/.

mary-poppins commented:
You can preview 553358b at https://pr43668-553358b.ngbuilds.io/.

IgorMinar commented on aio/content/guide/deprecations.md:
```diff
@@ -32,36 +32,38 @@ v11 - v14
-| `@angular/forms`        | [`FormBuilder.group` legacy options parameter](api/forms/FormBuilder#group)                   | <!--v11--> v14        |
-| `@angular/router`       | [`ActivatedRoute` params and `queryParams` properties](#activatedroute-props)                 | unspecified           |
-| template syntax         | [`/deep/`, `>>>`, and `::ng-deep`](#deep-component-style-selector)                            | <!--v7--> unspecified |
+| Area                      | API or Feature                                                                                | May be removed in     |
```
why such big diff? ws?

gkalpak replied:
Yeah. Only the two `@angular/service-worker` lines (near the bottom) were added. Everything else is whitespace to adjust cell sizes so the table remains aligned.

IgorMinar replied:
aha. thanks.

IgorMinar commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -37,7 +37,8 @@ Do this with the `checkForUpdate()` method:
 
-This method returns a `Promise` which indicates that the update check completed successfully, though it does not indicate whether an update was discovered as a result of the check. Even if one is found, the service worker must still successfully download the changed files, which can fail. If successful, the `available` event indicates availability of a new version of the application.
+This method returns a `Promise<boolean>` which indicates if an update is available for activation. The check might fail
+which will cause a rejection of the `Promise`.
```
one sentence per line please

IgorMinar reviewed (approved):
LGTM after cleanup. Thanks!

Reviewed-for: public-api, docs-packaging-and-releasing, fw-service-worker

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

mary-poppins commented:
You can preview 1bf7e72 at https://pr43668-1bf7e72.ngbuilds.io/.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api

IgorMinar commented:
I overrode the CLA label since the original PR has a CLA.

dylhunn commented:
This PR currently has merge conflicts with master; mind rebasing?

mary-poppins commented:
You can preview fd71566 at https://pr43668-fd71566.ngbuilds.io/.

gkalpak commented:
The CI failures are not related to this PR and will be fixed by #43714.

spike-rabbit committed:
feat(service-worker): expose more version update events

This commit introduces a new `SwUpdate#versionUpdates` observalbe, which provides more granular
information about Service Worker version updates than the already existing `SwUpdate#available`
observable.

The new events emitted by `SwUpdate#versionUpdates` basically add the possibility for application to
get notified if a new version is available on the server (not yet downloaded) and if an installation
of a new version failed.

Closes #39840

DEPRECATED: The `SwUpdate#availalbe` observable is deprecated.

The new `SwUpdate#versionUpdates` observable provides the same information and more. Therefore, it
is possible to rebuild the same behavior as `SwUpdate#availalbe` using the events emitted by
`SwUpdate#versionUpdates` and filtering for `VersionReadyEvent` events.
As a result, the `SwUpdate#availalbe` observable is now redundant.

gkalpak committed:
fixup! feat(service-worker): `SwUpdate#activeUpdate` and `SwUpdate#checkForUpdate` should have a meaningful outcome

gkalpak committed:
fixup! feat(service-worker): expose more version update events

gkalpak committed:
fixup! feat(service-worker): `SwUpdate#activeUpdate` and `SwUpdate#checkForUpdate` should have a meaningful outcome

spike-rabbit committed:
feat(service-worker): `SwUpdate#activeUpdate` and `SwUpdate#checkForUpdate` should have a meaningful outcome

Previously `activateUpdate` and `checkForUpdate` returned `Promise<void>`, which did not provide any
information of the actual outcome of the operation. Developers needed to subscribe to a
corresponding observable to obtain the outcome.

With this commit `SwUpdate#activateUpdate` will return `Promise<boolean>` which resolves to `true` if
an update was activated, to `false` if no new version is available and rejects in case of any error.
`SwUpdate#checkForUpdate` also returns a `Promise<boolean>` which resolves to `true` if an update is available, to `false` if no update was found and rejects in case of any error.

Closes #39840

BREAKING CHANGE: The return type of `SwUpdate#activateUpdate` and `SwUpdate#checkForUpdate` changed to `Promise<boolean>`.

Although unlikely, it is possible that this change will cause TypeScript type-checking to fail in
some cases. If necessary, update your types to account for the new
return type.

DEPRECATED: The `SwUpdate#activated` observable is deprecated.

The `SwUpdate#activated` observable only emits values as a direct response to calling
`SwUpdate#activateUpdate()` and was only useful for determining whether the call resulted in an
update or not. Now, the return value of `SwUpdate#activateUpdate()` can be used to determine the
outcome of the operation and therefore using `SwUpdate#activated` does not offer any benefit.

mary-poppins commented:
You can preview 88988f3 at https://pr43668-88988f3.ngbuilds.io/.

gkalpak commented:
**merge-assistance:** The CLA check was overridden (see https://github.com/angular/angular/pull/43668#issuecomment-932610176), but I had to rebase on master and now the CLA check fails again 😞

dylhunn commented:
Setting CLA to yes, since the original PR had CLA

dylhunn commented:
This PR was merged into the repository by commit 0dc45446fe487febaefaf68a928c5a249880f2f3.
